src/core/classes/page-load-watcher.ts

```diff
@@ -0,0 +1,157 @@
+        return this.staleWhileRevalidate;
+    }
+
+    /**
+     * Whether this load watcher had meaningful changes received in background.
+     *
+     * @return Whether this load watcher had meaningful changes received in background.
+     */
+    hasMeaningfulChanges(): boolean {
+        return this.hasChanges;
+    }
+
+    /**
+     * Set has meaningful changes to true.
+     */
+    markMeaningfulChanges(): void {
+        this.hasChanges = true;
+    }
+
+    /**
+     * Watch a component, waiting for it to be ready.
+     *
+     * @param component Component instance.
+     */
+    async watchComponent(component: AsyncComponent): Promise<void> {
+        this.components.add(component);
+        clearTimeout(this.finishTimeout);
+
+        try {
+            await component.ready();
+        } finally {
+            this.components.delete(component);
+            this.checkHasLoaded();
+        }
+    }
+
+    /**
+     * Get the reading strategy to use.
+     *
+     * @return Reading strategy to use.
+     */
+    getReadingStrategy(): CoreSitesReadingStrategy | undefined {
+        return this.staleWhileRevalidate ? CoreSitesReadingStrategy.STALE_WHILE_REVALIDATE : undefined;
+    }
+
+    /**
+     * Watch a WS request, handling the different values it can return, calling the hasMeaningfulChanges callback if needed to
+     * detect if there are new meaningful changes in the page load, and completing the page load when all requests have
+     * finished and all components are ready.
+     *
+     * @param observable Observable of the request.
+     * @param hasMeaningfulChanges Callback to check if there are meaningful changes if data was updated in background.
+     * @return First value of the observable.
+     */
+    watchRequest<T>(
+        observable: WSObservable<T>,
+        hasMeaningfulChanges?: (previousValue: T, newValue: T) => boolean,
+    ): Promise<T> {
+        return new Promise((resolve, reject) => {
+            let subscription: Subscription | null = null;
+            let firstValue: T | undefined;
+            this.ongoingRequests++;
+            clearTimeout(this.finishTimeout);
+
+            const complete = async () => {
+                this.ongoingRequests--;
+                this.checkHasLoaded();
+
+                // Subscription variable might not be set because the observable completed immediately. Wait for next tick.
+                await CoreUtils.nextTick();
+                subscription?.unsubscribe();
+            };
+
+            subscription = observable.subscribe({
+                next: value => {
+                    if (!firstValue) {
+                        firstValue = value;
+                        resolve(value);
+
+                        return;
+                    }
+
+                    // Second value, it means data was updated in background. Compare data.
+                    if (hasMeaningfulChanges && hasMeaningfulChanges(firstValue, value)) {
```

Comment:
```suggestion
                    if (hasMeaningfulChanges?.(firstValue, value)) {
```

I don't remember if I already suggested this, sorry if I did. But in case you didn't know, this is how you use optional chaining with method calls :). This can be also applied to other parts of this PR, and the codebase overall if we like it (I do?).

Reply:
I think I've used it in some places, but IIRC some other dev in the team (I can't remember who it was, maybe Juan) said it looked a bit weird so I stopped using it. I'm ok with using it so I changed it, maybe we can talk about this in one of our dev meetings.